challenges/systems/challenge-2.md

```diff
@@ -0,0 +1,294 @@
+# Summary
+
+Implement a prototype load balancer.
+
+# Rationale
+
+This exercise has two goals:
+
+* It helps us to understand what to expect from you as a developer, how you
+  write production code, how you reason about API design and how you
+  communicate when trying to understand a problem before you solve it.
+* It helps you get a feel for what it would be like to work at Teleport, as
+  this exercise aims to simulate our day-as-usual and expose you to the type of
+  work we're doing here.
+
+We believe this technique is not only better, but also is more fun compared to
+whiteboard/quiz interviews so common in the industry. It's not without the
+downsides - it could take longer than traditional interviews.
+
+[Some of the best teams use coding challenges.](https://sockpuppet.org/blog/2015/03/06/the-hiring-post/)
+
+We appreciate your time and are looking forward to hack on this project
+together.
+
+# Levels
+
+There are 6 engineering levels at Teleport. It's possible to score on level 1-5
+through coding challenge.
+
+Level 6 is only for internal promotions. Check the [engineering levels
+document](https://raw.githubusercontent.com/gravitational/careers/main/levels.pdf)
+for more details.
+
+# Interview process
+
+The interview process will start with you receiving an invite to a private
+Slack channel. That channel will contain the interview panel. You can ask them
+about the engineering culture, work and life balance, or anything else that you
+would like to learn about Teleport.
+
+Before writing any actual code, we ask that you write a brief design document.
+The design document should cover: design approach, scope, proposed API,
+security considerations, CLI UX, and implementation details where appropriate.
+
+Please submit the design document as Markdown and PR#1 to your repository to
+allow us to provide you feedback on the proposed design.
+
+A few notes about the design document:
+
+* We expect the design document to be complete roughly within the first week.
+  This is to ensure you have enough time to work on the implementation.
+* Avoid writing an overly detailed design document. Two to three pages is
+  sufficient.
+* Avoid sending us draft design documents, it is difficult to evaluate which
+  parts are draft and which parts are complete. Instead we encourage asking
+  questions in Slack and sharing a design document that is ready to be reviewed.
+
+Once the design document has been approved by two reviewers, move onto
+implementation.
+
+Split your code submission into roughly 3-5 Pull Requests to give the team an
```

Comment:
Should we update the guidance here? We said the design document is roughly a week, so if candidates follow that guidance they have only the 2nd week (roughly 5 work days) to put together 3-5 PRs. At the high end that's almost a PR a day, which may be hard for reviewers to keep up with.

Reply:
Good point, what you do think about reducing the recommended time for the design document?

Reply:
That lines up with what I tell candidates. I usually say 2-3 days on the design doc is a good rule of thumb.